src/material/button/_button-base.scss

```diff
@@ -133,3 +133,48 @@
     }
   }
 }
+
+@mixin mat-private-button-horizontal-layout($prefix, $slots, $has-with-icon-padding) {
+  @include token-utils.use-tokens($prefix, $slots) {
+    $icon-spacing: token-utils.get-token-variable-reference(icon-spacing, true);
+    $icon-offset: token-utils.get-token-variable-reference(icon-offset, true);
+    $horizontal-padding: token-utils.get-token-variable-reference(horizontal-padding, true);
+    padding: 0 $horizontal-padding;
+
+    @if ($has-with-icon-padding) {
+      $with-icon-horizontal-padding:
+        token-utils.get-token-variable-reference(with-icon-horizontal-padding, true);
+
+      // stylelint-disable-next-line selector-class-pattern
+      &:has(.material-icons, mat-icon, [matButtonIcon]) {
+        padding: 0 $with-icon-horizontal-padding;
+      }
+    }
+
+    // MDC expects button icons to contain this HTML content:
+    // ```html
+    //   <span class="mdc-button__icon material-icons">favorite</span>
+    // ```
+    // However, Angular Material expects a `mat-icon` instead. The following
+    // styles will lay out the icons appropriately.
+    & > .mat-icon {
```

Comment:
Feel free to not do this if it makes landing the change more difficult, but we should be able to use `margin-inline-start` and `margin-inline-end` now, which could simplify this type of RTL logic

Reply:
Same reason as below: I tried keeping the declarations as close to MDC as possible.